compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1143,6 +1143,57 @@ fn link_sanitizer_runtime(sess: &Session, linker: &mut dyn Linker, name: &str) {
     }
 }
 
+fn add_profiler_libraries(sess: &Session, crate_type: CrateType, linker: &mut dyn Linker) {
+    let needs_runtime = match crate_type {
+        CrateType::Executable => true,
+        CrateType::Dylib
+        | CrateType::Cdylib
+        | CrateType::ProcMacro
+        | CrateType::Rlib
+        | CrateType::Staticlib => false,
+    };
+
+    if !needs_runtime {
+        return;
+    }
+
+    if !sess.opts.unstable_opts.no_profiler_runtime
+        && (sess.instrument_coverage()
+            || sess.opts.unstable_opts.profile
+            || sess.opts.cg.profile_generate.enabled())
+        // If user doesn't provide custom profiler runtime, link default llvm profiler.
+        && sess.opts.unstable_opts.profiler_runtime.is_none()
+    {
+        link_profiler_runtime(sess, linker);
+    }
+}
+
+fn link_profiler_runtime(sess: &Session, linker: &mut dyn Linker) {
+    fn find_profiler_runtime(sess: &Session, filename: &str) -> PathBuf {
+        let session_tlib =
+            filesearch::make_target_lib_path(&sess.sysroot, sess.opts.target_triple.triple());
+        let path = session_tlib.join(filename);
+        if path.exists() {
+            return session_tlib;
+        } else {
+            let default_sysroot = filesearch::get_or_default_sysroot();
+            let default_tlib = filesearch::make_target_lib_path(
+                &default_sysroot,
+                sess.opts.target_triple.triple(),
+            );
+            return default_tlib;
```

Comment:
Why are we using the default sysroot if the user has explicitly passed `--sysroot`? This doesn't look correct.

Reply:
This is the fallback, which is inspired by https://github.com/rust-lang/rust/pull/80933/files

Reply:
The sysroot is honored in the above branch.

Reply:
I merged this function with old code to make it easier to review.